### PR TITLE
Add material request route

### PR DESCRIPTION
**Pull request by Philinphiladelphia**
## Issue Number

#146 

## Purpose/Implementation Notes

Adds routes and tests for material-request

## Types of changes

- New feature (non-breaking change which adds functionality)

## Functional tests

Unit tests

## Checklist

- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: api/resources_portal/migrations/0001_initial.py [hunk 1]
```diff
@@ -734,6 +742,25 @@ class Migration(migrations.Migration):
...
+            name="groups",
+            field=models.ManyToManyField(
+                blank=True,
+                help_text="The groups this user belongs to. A user will get all permissions granted to each of their groups.",
+                related_name="user_set",
+                related_query_name="user",
+                to="auth.Group",
+                verbose_name="groups",
+            ),
+        ),
```
Comment (kurtwheeler) [inline api/resources_portal/migrations/0001_initial.py:763]: Hey so we're actually at the point where I think we should switch back to adding new migrations, because we have the staging server up and I don't really want to have to re-init its database :)

Comment (Philinphiladelphia) [reply api/resources_portal/migrations/0001_initial.py]: Gotcha. I just separated it out (:

File: api/resources_portal/migrations/0001_initial.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
-# Generated by Django 2.2.12 on 2020-06-01 17:24
+# Generated by Django 2.2.12 on 2020-05-08 15:15
```
Comment (kurtwheeler) [inline api/resources_portal/migrations/0001_initial.py:1]: I'm confused, why is this file still changing?

File: api/resources_portal/test/views/__init__.py [hunk 3]
```diff
@@ -0,0 +1,4 @@
+from resources_portal.test.views.test_material_request import (
+    TestMaterialRequestListTestCase,
+    TestSingleMaterialRequestTestCase,
+)
```
Comment (kurtwheeler) [inline api/resources_portal/test/views/__init__.py:4]: Why is this necessary?

File: api/resources_portal/views/material_request.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+            "created_at",
+            "updated_at",
+        )
+        read_only_fields = ("id", "created_at", "updated_at")
+
+
+class MaterialRequestDetailSerializer(MaterialRequestSerializer):
+    assigned_to = UserSerializer()
+    requester = UserSerializer()
+    material = MaterialSerializer()
```
Comment (kurtwheeler) [inline api/resources_portal/views/material_request.py:38]: Can you actually use the *RelationSerializer for users and materials as well? The non-relation serializers have nested attributes, which we don't want. For example I'd expect a response from this endpoint to have a user object for who it's assigned to, but I don't need to see every org that user belongs to.

File: api/resources_portal/test/views/__init__.py (hunk 3 above)
Comment (Philinphiladelphia) [reply api/resources_portal/test/views/__init__.py:4]: Ah- this was my mistake. I had it here so I could run these tests specifically instead of all of them. I'll remove.

Comment (kurtwheeler) [reply api/resources_portal/test/views/__init__.py:4]: Ohhhh. You couldn't run just them without it?

Comment (Philinphiladelphia) [reply api/resources_portal/test/views/__init__.py:4]: I don't think so. In order to run specific tests, you need to make the module which contains the tests visible to the run_tests.sh script. At some point I'll make a PR which adds all the tests to the __init__.py so the script can see them, but it's out of scope for this.

File: api/resources_portal/models/material_request.py [hunk 5]
```diff
@@ -12,6 +12,13 @@ class Meta:
 
     objects = models.Manager()
 
+    STATUS_CHOICES = (
+        ("PENDING", "PENDING"),
+        ("APPROVED", "APPROVED"),
+        ("REJECTED", "REJECTED"),
+        ("INVALID", "INVALID"),
```
Comment (kurtwheeler) [inline api/resources_portal/models/material_request.py:19]: It just occurred to me that we should also have a `CANCELLED` state for if the requester cancels the request. This is different than `REJECTED` in that it was cancelled by the requester, rather than the sharer.

File: api/resources_portal/views/material_request.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+                CanViewRequests,
+            ]
+        elif self.action == "retrieve":
+            permission_classes = [IsAuthenticated, CanViewRequestsOrIsRequester]
+        elif self.action == "update" or self.action == "partial-update":
+            permission_classes = [IsAuthenticated, CanApproveRequestsOrIsRequester]
+        elif self.action == "destroy":
+            permission_classes = [
+                IsAuthenticated,
+                IsRequester,
```
Comment (kurtwheeler) [inline api/resources_portal/views/material_request.py:97]: Hmm, so it looks like currently we're just allowing the user to delete rather than cancel their requests.

@davidsmejia do you think allowing these to be deleted is good or should we add a `CANCELLED` state that the requester can update the request to? I'm leaning towards having `CANCELLED` but I could maybe be convinced otherwise.

Comment (davidsmejia) [reply api/resources_portal/views/material_request.py:97]: I agree with your leanings. We should show cancelled for historical and reporting purposes

File: api/resources_portal/migrations/0002_auto_20200604_2003.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+                    ("APPROVED", "APPROVED"),
+                    ("REJECTED", "REJECTED"),
+                    ("INVALID", "INVALID"),
+                    ("CANCELLED", "CANCELLED"),
+                ],
+                default="PENDING",
+                max_length=32,
+            ),
+        ),
+    ]
```
Comment (kurtwheeler) [inline api/resources_portal/migrations/0002_auto_20200604_2003.py:28]: Wait, can this one be rolled up into the initial migration? (Sorry if I've been confusing about this. Basically once we have a staging server up I want to start adding additional migrations, but let's start that after this PR since I gotta re-init the staging database anyway.

File: api/resources_portal/views/material_request.py [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+                material_request.requester_signed_mta_attachment = serializer.validated_data[
+                    "requester_signed_mta_attachment"
+                ]
+            if "status" in request.data and request.data["status"] != material_request.status:
+                if serializer.validated_data["status"] != "CANCELLED":
+                    return Response(status=403)
+                else:
+                    material_request.status = serializer.validated_data["status"]
+        else:
+            if "status" in request.data:
```
Comment (kurtwheeler) [inline api/resources_portal/views/material_request.py:164]: Can we also check that `request.data["status"] != "CANCELLED"`? I'd like for that state to only be set by the requester, otherwise the request should be `REJECTED`.

Comment (Philinphiladelphia) [reply api/resources_portal/views/material_request.py:167]: sure! done.
